src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1476,5 +1478,182 @@ public static double ScaleB(double x, int n)
             double u = BitConverter.Int64BitsToDouble(((long)(0x3ff + n) << 52));
             return y * u;
         }
+
+        //
+        // Helpers, those methods are referenced from the JIT
+        //
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint High32Bits(ulong a) => (uint)(a >> 32);
+
+        [StackTraceHidden]
+        internal static long MultiplyInt64Checked(long left, long right)
+        {
```

Comment:
This is copy&paste from MathHelpers.cs with no changes.

Reply:
This seems like something we could simplify long term.

It should be possible to implement this like `Int128 operator checked *` is, in terms of `BigMul(long, long)` and simply checking if the upper bits match the sign of the msb of low.

The JIT should already be handling `uint * uint = ulong` which should keep it efficient and simple on 32-bit platforms (since most provide some built-in way to do `32 x 32 = 64`).

Reply:
> Int128 operator checked * is, in terms of BigMul(long, long) and simply checking if the upper bits match the sign of the msb of low.

I think `Int128 operator checked *` is leaving perf on the table. It does 4 multiplications, but it only needs to do 3 multiplications. The result of 4th multiplication is only used to perform an overflow check. I would expect that checking the operands directly is going to be cheaper. 

I do not see an obvious opportunity here, but I do not mind if somebody wants to spend time tweaking this helper and proving that the new implementation has better performance for a typical mix of inputs.